src/veil/ringct/anonwallet.cpp

```diff
@@ -1565,7 +1566,12 @@ int AnonWallet::AddCTData(CTxOutBase *txout, CTempRecipient &r, std::string &sEr
     size_t nRangeProofLen = 5134;
     pvRangeproof->resize(nRangeProofLen);
 
+    //If this is a proof-of-stake transaction, then the minimum value must be set above 0
     uint64_t min_value = 0;
+    if (fProofOfStake) {
+        min_value = GetStakeWeightBracket(nValue);
```

Comment:
Should be 10.

Reply:
You have to use something that the public will be able to validate. As they won't know the actual value, they will only have the min and max values that are in the rangeproof. For obvious reasons you would want to round down to the min value.

Reply:
We got a solution for this, in fact, we are limited by how Rangeproofs are created. It's impossible to set a maximum value to be anything, it must be an even power of 2 (2^2, 2^4, 2^6 etc..). So, you can commit to a bottom value + 1 satoshi (so people can not stake in multiple denom brackets). The only real big change that kind of threw us for a loop was when we found out that someone can stake every bracket denomination. So, the only may to set a max is to actually limit the size of the rangeproof itself.

Through discussions with people that are more familiar with the libsecp256k1-zkp library, that's when we realised that all of the denomination and weights need to change entirely. We also realised that through some simple math we are able to quickly calculate what is the min max denom bracket values without making them static and having it apply to across the whole range of coins.

I wrote up some simulation code that is able to take in all the parameters that we are using and able to find which brackets are more unfair to others as the goal is to make them as equal as possible. You can see and play around with it on my github if you are curious. veil_pos_simulation

Reply:
@presstab thanks for jumping in.   as @mimirmim touched on; the min/max ranges need to be defined by the number of bits that will define the private portion of a TXO value.   The number of privacy bits will increase as the TXO value increases (by a logarithmic function).  With that said, as we get into smaller denoms, the privacy bits will approach 1.  This actually allows us to stake TXO dust.  So there really isn't any need for a true minimum defined in the code.  The algorithm defines the min and max that a TXO will fall into, and consensus can confirm that the staking transaction confirms to a min and max [ `(min-1sat)*4 == max` ].  The TXO value is within that range and the staking weight is the min.

I'm thinking of having a user configurable min_value.  Let them define (or override) the default min_value to whatever dust size they don't want to have considered for staking; that way they know what TXOs will be considered, and won't be considered.

Reply:
I don't see any reason that a max value has any importance when it comes to bracketing/stake weights. The weight does not change the proof hash, so focusing on max value would not be a measure to prevent grinding. 

Being able to "stake every bracket denomination" also doesn't increase the chance of finding a successful stake. 
The formula for a winning hash is roughly:
`success = hash(timestamp + Immutable UTXO Data) < (difficulty * weight)`
If a stake is valid under one bracket, mathematically it is also valid under all bigger bracket levels as well. However using a bigger bracket would provide less reward, so would be against the interest of the staker and no real reason to do (unless looking for some additional obfuscation).

@CaveSpectre11 you don't want to be able to stake dust. At that point you would call it proof of work, not proof of stake.

Reply:
But, vice versa, if you have 10,000 and the `difficulty * weight` can have a chance at still being true at lower denominations, then I could just say that my 10,000 weight is a lower weight, getting a better and unfair reward. You only have you prove that are above a minimum value that you committed to.

A big advantage to limiting the rangeproof size is its "cheaper" to produce smaller denom hashes since the size is reduced. Also, a bit less bloat on the blockchain which I consider quite a large win seeing as our blockchain size is exploding right now.

Reply:
Having smaller proofs that are less computational is a great thing. Still does not change the mathematics that a valid stake hash of a small amount of coins will still be just as valid with a larger max value bound. There isn't any real problem with that either, just saying that the important bound is the min not the max.

> We have discussed and will implement a minimum default value to ensure that nobody will be accidentally wasting their time staking denominations that won't hit a reward in years. 

I think you are missing the point. I could split a few thousand Veil into hundred of millions of dust UTXOs and then use a GPU to grind through all of those outputs and find a valid stake quicker than anyone else would be able to. If a system is designed to allow that, then that system is called proof of work.

Reply:
I do get that and that is one of the biggest issues with PoS that you can split up your balances into many smaller UTXOs to have more permutations which would turn it into a PoW-like system (and something that I talk about a lot). This issue even exists no matter where the floor is provided with a large enough balance. A minimum value that you must have certainly can discourage this. However, that does bring in now a cost to actually create those coins.

i.e what's better? Rolling a many-sided die and having your result modified to provide you with a win if you roll a high enough number, or rolling that die many times as fast as you can until you get the high enough number. I've seen lots of people say the latter than the former and they may be right but, I need to see those results to be able to actually conclusively agree with that.

I will create a sim in practice where I write 2 players does exactly that except with an RNG instead of hashing. One with a large balance which his weight * difficulty, or one with many chances. May seem like a waste of time but its always better to come out with proof for decisions.

Reply:
So, one important observation that I had in the simulation is that those staking with 10,000 denominations didn't have 30% less, they had between 50% to 35% less after 10 years of staking. That suggests that even with staking all 10 denominations they have a clear advantage and the higher amounts should -not- be punished.

@paulmoonshine and I just worked out the math if someone with more permutations had an advantage and probabilistically, they certainly do. So, those that are staking at smaller denominations should be punished to flatten their clear advantage.